src/Raven.Client/Documents/BulkInsert/BulkInsertOperation.cs

```diff
@@ -423,11 +506,13 @@ private async Task HandleErrors(string documentId, Exception e)
             await ThrowOnUnavailableStream(documentId, e).ConfigureAwait(false);
         }
 
-        private IDisposable ConcurrencyCheck()
+        private async Task<IDisposable> ConcurrencyCheck()
```

Comment:
add Async

Reply:
Let's do it a bit differently

1. return `ValueTask<ReleaseStreamLock>` 
2. implement readonly struct that implements IDisposable and releases the semaphore
3. use it everywhere with 'using'
4. remove all the '_concurrentCheck' code and substitute this with `.WaitAsync(0)` on semaphore. If this will return false throw the cannot be executed concurrently exception

Reply:
but we don't want to throw when we have a regular write and the semaphore is taken by the heartbeat write.
we need some lock on the stream and another check for the concurrency to know where to throw

Reply:
In this case, RWLS would be a better option.